yt/utilities/answer_testing/testing_utilities.py

```diff
@@ -0,0 +1,557 @@
+    ds.index
+    return ds
+
+
+def requires_ds(ds_fn, file_check=False):
+    r"""
+    Meta-wrapper for specifying required data for a test and
+    checking if said data exists.
+    """
+
+    def ffalse(func):
+        @functools.wraps(func)
+        def skip(*args, **kwargs):
+            msg = f"{ds_fn} not found, skipping {func.__name__}."
+            pytest.skip(msg)
+
+        return skip
+
+    def ftrue(func):
+        @functools.wraps(func)
+        def true_wrapper(*args, **kwargs):
+            return func(*args, **kwargs)
+
+        return true_wrapper
+
+    if not can_run_ds(ds_fn, file_check):
+        return ffalse
+    else:
+        return ftrue
+
+
+def requires_sim(sim_fn, sim_type, file_check=False):
+    r"""
+    Meta-wrapper for specifying a required simulation for a test and
+    checking if said simulation exists.
+    """
+
+    def ffalse(func):
+        @functools.wraps(func)
+        def skip(*args, **kwargs):
+            msg = f"{sim_fn} not found, skipping {func.__name__}."
+            pytest.skip(msg)
+
+        return skip
+
+    def ftrue(func):
+        @functools.wraps(func)
+        def true_wrapper(*args, **kwargs):
+            return func
+
+        return true_wrapper
+
+    if not can_run_sim(sim_fn, sim_type, file_check):
+        return ffalse
+    else:
+        return ftrue
+
+
+def create_obj(ds, obj_type):
+    # obj_type should be tuple of
+    #  ( obj_name, ( args ) )
+    if obj_type is None:
+        return ds.all_data()
+    cls = getattr(ds, obj_type[0])
+    obj = cls(*obj_type[1])
+    return obj
+
+
+def compare_unit_attributes(ds1, ds2):
+    r"""
+    Checks to make sure that the length, mass, time, velocity, and
+    magnetic units are the same for two different dataset objects.
+    """
+    attrs = ("length_unit", "mass_unit", "time_unit", "velocity_unit", "magnetic_unit")
+    for attr in attrs:
+        u1 = getattr(ds1, attr, None)
+        u2 = getattr(ds2, attr, None)
+        assert u1 == u2
+
+
+def fake_halo_catalog(data):
+    filename = "catalog.0.h5"
+    ftypes = {field: "." for field in data}
+    extra_attrs = {"data_type": "halo_catalog", "num_halos": data["particle_mass"].size}
+    ds = {
+        "cosmological_simulation": 1,
+        "omega_lambda": 0.7,
+        "omega_matter": 0.3,
+        "hubble_constant": 0.7,
+        "current_redshift": 0,
+        "current_time": YTQuantity(1, "yr"),
+        "domain_left_edge": YTArray(np.zeros(3), "cm"),
+        "domain_right_edge": YTArray(np.ones(3), "cm"),
+    }
+    save_as_dataset(ds, filename, data, field_types=ftypes, extra_attrs=extra_attrs)
+    return filename
+
+
+def _create_plot_window_attribute_plot(ds, ptype, field, axis, pkwargs=None):
+    r"""
+    Convenience function used in plot_window_attribute_test.
+
+    Parameters
+    ----------
+    ds : Dataset
+        The Dataset object from which the plotting data is taken.
+
+    ptype : string
+        Type of plot to make (e.g., SlicePlot).
```

Comment:
Everywhere else in the code, `ptype` is particle type. Shouldn't we rename that to `plot_type` as in `_create_phase_plot_attribute_plot`?

Reply:
Good catch!